### PR TITLE
login: Prevent multiple logins in a single browser session

### PR DESCRIPTION
**Pull request by mvollmer**
unless allowed by cockpit.conf.

If the login page is loaded and a valid session cookie is already available, then we are about to log into a second host from the same browser session, and both logins will have access to each others cookies. This should only be allowed when AllowMultiHost is true. If it is not true, the login page immediately redirects to the session for the existing cookie.

Information about session cookies is not available to login.js, so cockpit-ws helps out by exposing which ones are present, without exposing the cookies themselves.

- [x] tests
- [x] coverage

### CODE REVIEW

**Comment by mvollmer** [comment]
This is the first half of #20834. I think we should get it in, before we figure out the warning texts.

**Comment by mvollmer** [comment]
Nice,  test/verify/check-embed TestEmbed.testBasic got broken, so we do test this!

File: pkg/static/login.js [hunk 1]
```diff
@@ -338,6 +338,26 @@ import "./login.scss";
         event.stopPropagation();
     }
 
+    function deal_with_multihost() {
+        // If we are currently logged in to some machine, but still
+        // end up on the login page, we are about to load resources
+        // from two machines into the same browser origin. This needs
+        // to be allowed explicitly via a configuration setting.
+
+        const logged_into = environment.logged_into || [ ];
```
Comment (cockpituous) [inline pkg/static/login.js:347]: This added line is not executed by any test.

File: pkg/static/login.js [hunk 2]
```diff
@@ -338,6 +338,26 @@
...
+        const cur_machine = logged_into.length > 0 ? logged_into[0] : null;
+
+        function redirect_to_current_machine() {
+            if (cur_machine === ".")
+                login_reload("/");
```
Comment (cockpituous) [inline pkg/static/login.js:352]: These 3 added lines are not executed by any test.

File: pkg/static/login.js [hunk 3]
```diff
@@ -338,6 +338,26 @@
...
+            else
+                login_reload("/=" + cur_machine);
```
Comment (cockpituous) [inline pkg/static/login.js:354]: This added line is not executed by any test.

File: pkg/static/login.js [hunk 4]
```diff
@@ -338,6 +338,26 @@
...
+        }
+
+        if (cur_machine && !environment.page.allow_multihost)
+            redirect_to_current_machine();
```
Comment (cockpituous) [inline pkg/static/login.js:358]: This added line is not executed by any test.

File: pkg/static/login.js [hunk 5]
```diff
@@ -338,6 +338,26 @@ import "./login.scss";
         event.stopPropagation();
     }
 
+    function deal_with_multihost() {
+        // If we are currently logged in to some machine, but still
+        // end up on the login page, we are about to load resources
+        // from two machines into the same browser origin. This needs
+        // to be allowed explicitly via a configuration setting.
+
+        const logged_into = environment.logged_into || [];
```
Comment (cockpituous) [inline pkg/static/login.js:347]: This added line is not executed by any test.

File: pkg/static/login.js [hunk 6]
```diff
@@ -338,6 +338,26 @@ import "./login.scss";
...
+        const cur_machine = logged_into.length > 0 ? logged_into[0] : null;
+
+        function redirect_to_current_machine() {
+            if (cur_machine === ".")
+                login_reload("/");
```
Comment (cockpituous) [inline pkg/static/login.js:352]: These 3 added lines are not executed by any test.

File: pkg/static/login.js [hunk 7]
```diff
@@ -338,6 +338,26 @@ import "./login.scss";
...
+            else
+                login_reload("/=" + cur_machine);
```
Comment (cockpituous) [inline pkg/static/login.js:354]: This added line is not executed by any test.

File: pkg/static/login.js [hunk 8]
```diff
@@ -338,6 +338,26 @@ import "./login.scss";
...
+        }
+
+        if (cur_machine && !environment.page.allow_multihost)
+            redirect_to_current_machine();
```
Comment (cockpituous) [inline pkg/static/login.js:358]: This added line is not executed by any test.

**Comment by mvollmer** [comment]
Nicely testing this requires Browser tab support in testlib.py. Let's not block on that. The check-embed test covers this a bit, fortunately.

File: pkg/static/login.js (hunk 5 above)
Comment (cockpituous) [inline pkg/static/login.js:347]: This added line is not executed by any test.

File: pkg/static/login.js (hunk 6 above)
Comment (cockpituous) [inline pkg/static/login.js:352]: This added line is not executed by any test.

File: pkg/static/login.js (hunk 7 above)
Comment (cockpituous) [inline pkg/static/login.js:354]: This added line is not executed by any test.

File: src/ws/cockpitauth.c [hunk 9]
```diff
@@ -1241,13 +1241,38 @@ base64_decode_string (const char *enc)
...
+{
+  gchar *decoded = NULL;
+  const char *prefix = "v=2;k=";
+  CockpitSession *ret = NULL;
+
+  g_return_val_if_fail (self != NULL, FALSE);
+
+  if (cookie_value)
+    {
+      decoded = base64_decode_string (cookie_value);
```
Comment (martinpitt) [inline src/ws/cockpitauth.c:1256]: You can use `g_autofree` here, drop the declaration in line 1248 and the g_free below.

**update**: Oh, this is moved code. OK to keep if you prefer not to touch it.

File: pkg/static/login.js [hunk 10]
```diff
@@ -338,6 +338,27 @@ import "./login.scss";
...
     }
 
+    function deal_with_multihost() {
+        // If we are currently logged in to some machine, but still
+        // end up on the login page, we are about to load resources
+        // from two machines into the same browser origin. This needs
+        // to be allowed explicitly via a configuration setting.
+
+        const logged_into = environment.logged_into || [];
+        const cur_machine = logged_into.length > 0 ? logged_into[0] : null;
```
Comment (martinpitt) [inline pkg/static/login.js:348]: `const cur_machine = environment.logged_into?.[0]`

(but if you prefer this, it's fine)

File: src/ws/cockpitauth.h [hunk 11]
```diff
@@ -102,7 +102,10 @@ gboolean        cockpit_auth_local_finish    (CockpitAuth *self,
 CockpitWebService *  cockpit_auth_check_cookie    (CockpitAuth *self,
                                                    CockpitWebRequest *request);
 
-gchar *         cockpit_auth_parse_application    (const gchar *path,
+gboolean        cockpit_auth_is_valid_cookie_value (CockpitAuth *self,
+                                                    const gchar *cookie_value);
+
+  gchar *         cockpit_auth_parse_application    (const gchar *path,
```
Comment (martinpitt) [inline src/ws/cockpitauth.h:108]: Please outdent again

File: src/ws/cockpithandlers.c [hunk 12]
```diff
@@ -277,8 +277,47 @@ add_page_to_environment (JsonObject *object,
...
+
+    if (*equal != '=')
+      continue;
+
+    g_autofree gchar *value = g_strndup (equal + 1, end - equal - 1);
+
+    if (!cockpit_auth_is_valid_cookie_value (auth, value))
+      continue;
+
+    g_autofree gchar *name = g_strndup (start, equal - start);
```
Comment (martinpitt) [inline src/ws/cockpithandlers.c:309]: Hmm, do you want `sscanf()` here?

**Review by martinpitt — COMMENTED** [review]
Thanks! This wasn't on my radar (no review request). This by and large looks right, and it would indeed be a really nice fix for tomorrow's release. I found a few style issues, but they are not critical and could become follow-ups. Are you otherwise happy with this?

File: src/ws/cockpithandlers.c (hunk 12 above)
Comment (mvollmer) [reply src/ws/cockpithandlers.c:309]: Oh, that would be the second time in my life that I use sscanf... hmm.

File: pkg/static/login.js (hunk 10 above)
Comment (mvollmer) [reply pkg/static/login.js:348]: Looks too much like Perl for my taste... :-) but let's do it.

**Comment by martinpitt** [comment]
I tested this with the flatpak, and it behaves the same way: Opening a second window goes straight into the running session. We should be able to improve that by isolating the webkit instances from one another somehow, but that can be a follow-up.

**Review by martinpitt — APPROVED** [review]
Thanks! As discussed in chat, let's get that into today's release, and do some follow-ups.

**Pull request merged**